### ISSUE TITLE
API v2 (and legacy v1)

### ISSUE THREAD

**Issue by noplanman** [diasporg/Poduptime#82]
Time for an issue to discuss the legacy v1 and the new v2 of the API.

To make the development of the API easier and more logical, it would really be great to start implementing Slim for this, as routing can be made very easily.
Also, authentication via an API token can be easily added as a layer.

How will this be communicated to app developers and other API consumers though?

### v1
To keep the old API running for the moment, the easiest would be to have a view in the DB that still has all the "old" fields, mapped to the new ones.
Alternatively, there could be a second "pods" table which has the exact same structure, that gets updated from the new table periodically. The main advantage of this second approach, is that the new database structure can be changed without affecting the old API and at any given time when the v1 API goes offline, it can just be removed easily.

### v2
As we have a clean slate to work on here, let's come up with some nice ideas to populate this :blush:

**Comment by dmorley** [comment]
I think the idea of pods_old table is good, the current api can be easily moved to use that and legacy V1 is done.

I think it might make sense to get v1 back to v1 and then look to merge develop into master before we hash out v2.

**Comment by dmorley** [comment]
This might have been what you were thinking of but:

SELECT .. date_created AS dateCreated,date_updated AS dateUpdated ... FROM pods';

In the original api.php for the ones we changed seems super easy to me.?

**Comment by noplanman** [comment]
@dmorley That's a good idea to get the correct table names 👍 

The only problem is the columns that don't exist any more in v2.
So I think the easiest would be to have a copy of the table and update it with a small script, filling dummy/predefined values into the fields that only exist in v1.

**Comment by dmorley** [comment]
Agree, can't seem to wrap my brain around how to do this

**Comment by noplanman** [comment]
Also on this topic, for the new API it would be nice to be able to choose which fields to return, to minimise the request load if only a certain field is necessary (instead of delivering the whole bulk.

This seems to be common practice for APIs, which makes total sense.

I think we focus on putting this into v2, not v1 any more.
You OK for leaving v1 "as is" and legacy, right?

Adding the `fields=[array,of,fields]` parameter (or something like that), we could actually also add a simple filtering system, but I don't know how best to do that. (Trying to not overthink this...)

As an example:
`api.podupti.me/v2/pods?fields=[domain, score]&filter=[not-sure-how]`

This would easily allow to just extract `diaspora` pods for example, not the whole list. At the moment it's not too bad, it's just a few hundred. But in the future when there might be 1000s...

**Comment by dmorley** [comment]
Yes for sure v1 is dead, we just need to update that sql file with our changes to support it, once V2 is done and stable we can decide a deprecation for v1 timeline.

Yes love the idea of V2 being something the user can customize for sure. If designed right it should allow the filter and fields to be anything one wants.

**Comment by dmorley** [comment]
Makes sense to me to start drafting this out, started at https://github.com/diasporg/Poduptime/wiki/%5BDraft%5D-API-V2 Makes sense to me to get it mapped out before coding it.

**Comment by noplanman** [comment]
👍 

Since this is the first time we'd be building a solid API, I think we should take this chance to learn something new!

This seems to be a great tool to help with API development:
https://apiblueprint.org

**Comment by noplanman** [comment]
Just to keep this noted somewhere, would it be an idea to use `statistics.json` as a fallback in case `nodeinfo` can't be accessed?

**Comment by dmorley** [comment]
I used statistics.json for about 6 months after things were moved to nodeinfo and its been about a year now since the change, seems overkill to support pods that are > 1 year old code.

**Comment by noplanman** [comment]
Ok, didn't know that 😊 
In that case, we'll need to add some kind of "pod-deprecation", which we already talked about at some point, to add a flag in the db.

Instead of just `deleted`, I think it would make sense to have an enum that would allow different entries, so that it's clear what state the pod in the db is in.

**Comment by noplanman** [comment]
We should have a list somewhere, of bigger public services that use the current API, to be able to get in touch with them and let them know when v2 is available:

- https://social.mbuto.me/thewall/

**Comment by dmorley** [comment]
I think trolling github would be pretty good also, from what I have seen of anyone using, they post their code on github. https://github.com/search?utf8=%E2%9C%93&q=podupti.me%2Fapi.php&type=Code&ref=searchresults

**Comment by noplanman** [comment]
Awesome! 😃
Didn't even think of that...
